apps/ai-image-generator/backend/src/transform-images.ts

```diff
@@ -0,0 +1,158 @@
+import * as nodeFetch from 'node-fetch';
+import { default as sharp } from 'sharp';
+import { areEqualColors, toRGBA, toSharp } from './utils';
+
+interface TransformedImages {
+  image: Buffer;
+  mask: Buffer;
+}
+
+export const ERASE_COLOR: sharp.RGBA = { r: 174, g: 193, b: 204 };
+export const MAX_SIDE = 1024;
+
+interface Dimensions {
+  width: number;
+  height: number;
+}
+
+export class ImageTransformer {
+  constructor(
+    private readonly sourceImageResponse: nodeFetch.Response,
+    private readonly maskImageResponse: nodeFetch.Response
+  ) {}
+
+  async execute(): Promise<TransformedImages> {
+    const { image, dimensions: sourceImageDimensions } = await this.normalizeSourceImage();
+    const { height: imageHeight, width: imageWidth } = sourceImageDimensions;
+    const mask = await this.normalizeMaskImage({
+      width: imageWidth,
+      height: imageHeight,
+      eraseColor: ERASE_COLOR,
+    });
+    return {
+      image,
+      mask,
+    };
+  }
+
+  private async normalizeSourceImage(): Promise<{
+    image: Buffer;
+    dimensions: Dimensions;
+  }> {
+    const sharpImage = toSharp(this.sourceImageResponse.body);
+    const initialMetadata = await sharpImage.metadata();
```

Comment:
There's a subtle reason why I separated this. I will add a comment since it's non obvious.

The reason is that for certain formats it's possible _we will not have a width and height from the initial metadata_.

After we force the file to a PNG in the following line, we are _guaranteed_ to have a width and height since it's part of the PNG file spec. So that's why we do the check later, when we konw the values will be present. (And we error if not).

Reply:
Note that I will still streamline this and remove the unneeded interim variable.

Reply:
Okay got it, makes sense now. 👍